### ISSUE TITLE
Support explain_prediction for xgboost

### ISSUE THREAD

**Issue by lopuhin** [TeamHG-Memex/eli5#114]
The idea is to do something similar to http://blog.datadive.net/interpreting-random-forests/ - which is used by treeinterpreter (https://github.com/andosa/treeinterpreter/pull/7/).

In order to do that, we need to get the path that was used for prediction. There is a `pred_leaf` argument for ``Booster.predict``: http://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.Booster.predict that outputs indices of leafs in all trees. Having this indices, we can recover full paths having the tree structure.
One way to get the tree structure is to use ``get_dump`` method http://xgboost.readthedocs.io/en/latest/python/python_api.html#xgboost.Booster.get_dump which accepts ``dump_format="json"`` (only in xgboost master so far).
I'm not sure if there is an easier way to get it? But it seems these two pieces is all we need for explain_prediction in eli5.

**Comment by lopuhin** [comment]
A drawback of using ``dump_format="json"`` is that it's not released, and installing xgboost from github master is not a one-liner. Probably it's easier to parse text xgboost dump instead?

**Comment by kmike** [comment]
I think it makes sense to do both - use json if it is available and parse text if it is not. XGBoost itself used to parse its text format: https://github.com/dmlc/xgboost/blob/6082184cd1ee831c2ac5e30394cd78b1714cd8c9/python-package/xgboost/core.py#L1088

**Comment by lopuhin** [comment]
Cool, thanks for the pointer! I just implemented a converter from text to json format (https://github.com/TeamHG-Memex/eli5/commit/91c230f8104bbb92b5a918a18c06fa63c1b801cd#diff-ae7c2a29f85d11827eab60b1435ae8f4R55), but their code looks more robust :)

**Comment by lopuhin** [comment]
I think that now explain_prediction is working correctly for xgboost in https://github.com/TeamHG-Memex/eli5/compare/xgboost-prediction branch, and it makes sense, at least for mostly categorical features, like text.
I fear that this approach will not work as well for numerical features, e.g. when it is important that some value lies in an interval. There is also a comment in the original blog post that shows another weakness, that it will fail for a simple XOR tree (weight will be zero for top variable). In practice with xgboost we might have trees with different orderings of variables, but then both variables will have zero weight.
It might be possible to fix that if we track positive and negative contributions of each variable separately: then for XOR each variable will be both a positive and a negative feature, and same for interval example.

Currently the branch also has a lot of hacks and code in wrong place.

**Comment by kmike** [comment]
Fixed by #117.
